### PR TITLE
RedisSinkCluster Hiding/Handling modes

### PR DESCRIPTION
**Pull request by rukai**
Closes https://github.com/shotover/shotover-proxy/issues/631

* I opted to expose this to the user via an enum in the destination field, the changes to the topology files should demonstrate how this is used.
   + A result of this approach is that you can no longer specify a list of default nodes beyond the initial node.
      + I'm not sure if having the handling destination node as the only source of other nodes makes shotover more or less robust. On one hand shotover is pretending to be that node so if it cant contact it then its better to fall over to avoid hiding the issue. On the other hand working in more cases is always better right?
      + A clear benefit is it makes configuration easier as you only need to provide the one true destination and then other peers are sourced from that.
   + It seems reasonable to me, but I dont feel strongly about this approach and would be interested to hear feedback or other ideas.
   + I feel like a pretty good reason to not do this approach would be if cluster hiding mode was a much more common use case than cluster handling.
      + in that case we would be better off having cluster hiding as the default and then having a `handling_destination: Option<String>` field that enables handling mode
* the test_script case is absent from the handling integration test because script no longer runs against every instance of the cluster, which means that EVALSHA (which is routed by key) is likely to be run against a redis instance without the script.
* The `Flusher` abstraction needed to be added to the integration tests because the handling test cases were failing due to FLUSHDB only running on the destination instance instead of every instance of the cluster.
   + I honestly really dislike this solution but I cant think of a better one.

### CODE REVIEW

File: shotover-proxy/tests/redis_int_tests/basic_driver_tests.rs
Comment (conorbros) [inline shotover-proxy/tests/redis_int_tests/basic_driver_tests.rs:1576]: Another option for this could be integrating it into `ShotoverManager` and then passing that around instead of the Redis connection. That way we could remove some duplicate functionality.

Would have to rework lots of `ShotoverManager` though, so I'll let you decide if you want to do that. Otherwise this is fine.

Comment (rukai) [reply shotover-proxy/tests/redis_int_tests/basic_driver_tests.rs:1576]: I'm trying to imagine what such an API would look like.

```rust
let shotover_manager = ShotoverManager::from_topology_file("example-configs/redis-multi/topology.yaml");
// Store connections internally within shotover_manager: 6379 is shotovers redis port and 1111, 1112, 1113 are the ports of the actual redis instances.
shotover_manager.redis_connections_async(6379, &[1111, 1112, 1113]).await; 
run_test_case(&shotover_manager);
    
fn run_test_case(shotover_manager: &ShotoverManager) {
   shotover_manager.flush();
   shotover_manager.shotover_redis_connection.set("foo", "bar");
   for connection in shotover_manager.direct_redis_connections {
      assert_eq!(connection.get("foo"), "bar");
   }
}
```

I do quite like the idea of this but I'm worried of the extra statefulness within ShotoverManager that this introduces.
I think the simplified and more flexible API would be worth it though?

Something this doesnt account for is secondary shotover connections which can be useful for some testcases.
Maybe we can just externally manage that connection for those cases?

Comment (conorbros) [reply shotover-proxy/tests/redis_int_tests/basic_driver_tests.rs:1576]: >secondary shotover connections

That could work in a similar way to the `shotover_redis_connection` field you showed in your example above. They could be created lazily, reusing the port we originally specified to connect to Shotover.
```rust
let shotover_manager = ShotoverManager::from_topology_file("example-configs/redis-multi/topology.yaml");
// Store connections internally within shotover_manager: 6379 is shotovers redis port and 1111, 1112, 1113 are the ports of the actual redis instances.
shotover_manager.redis_connections_async(6379, &[1111, 1112, 1113]).await; 
run_test_case(&shotover_manager);
    
fn run_test_case(shotover_manager: &ShotoverManager) {
    shotover_manager.flush();
    shotover_manager.shotover_redis_connection(0).unwrap().set("foo", "bar"); // get our first shotover connection with the index 0
    shotover_manager.shotover_redis_connection(1).unwrap().set("foo", "bar"); // get our second shotover connection with the index 1
   for connection in shotover_manager.direct_redis_connections {
      assert_eq!(connection.get("foo"), "bar");
   }
}
```

File: shotover-proxy/example-configs/redis-cluster-dr/topology.yaml
Comment (benbromhead) [inline shotover-proxy/example-configs/redis-cluster-dr/topology.yaml:28]: Hiding here as a keyword feels confusing (are we hiding these nodes or are we including these in the hidden cluster).

File: shotover-proxy/example-configs/redis-cluster-handling/topology.yaml
Comment (benbromhead) [inline shotover-proxy/example-configs/redis-cluster-handling/topology.yaml:12]: As above... I think destination (or something like it) needs to be the contact point keyword and Handling / Hiding modes defined above. Or even an explicity `mode: Handling` property.

I know this makes the rust side less ergonomic, but right now the user configuration feels confusing with the way it's set up.

File: shotover-proxy/src/transforms/redis/sink_cluster.rs
Comment (benbromhead) [inline shotover-proxy/src/transforms/redis/sink_cluster.rs:69]: call this mode

File: shotover-proxy/src/transforms/redis/sink_cluster.rs
Comment (benbromhead) [inline shotover-proxy/src/transforms/redis/sink_cluster.rs:71]: make the string a named field called destination or something

**Review by benbromhead — COMMENTED** [review]
Having only one contact/discovery points is problematic and likely to lead to the occasional discovery related outage (C* recommends bootstrapping nodes have 2 contact seeds and that's just on bootstrap which is a once off process) , I would err on the side of 2.

Could you also update the docs describing the Redis transform to be in line with this ability to support two different modes.

**Comment by rukai** [comment]
I changed the public facing API to use a single direct_connection field.
I think that introducing an enum here isnt worth the cost because enums can be confusing to configure in yaml.
And I think the hiding/handling distinction I invented is not clear by itself (the user would have to check the documentation anyway)
So we may as well just use an Option that changes functionality when set and document this functionality in the config example.

But I guess you could also argue that having a mandatory Hiding/Handling enum ensures that users always know that such a configuration is active rather than hiding it behind a default value?

I'll look into what we want to do with the possible ShotoverManager refactor.

File: shotover-proxy/tests/redis_int_tests/basic_driver_tests.rs
Comment (rukai) [reply shotover-proxy/tests/redis_int_tests/basic_driver_tests.rs:1576]: After thinking about it I would like to land the Flusher as is and raise an issue to investigate combining with ShotoverManager.

My reasoning is:
It will touch a lot of things and so is out of scope of this PR and would therefore need to be a prereq or postreq. Making it a postreq has the advantage of allowing us to directly compare the ShotoverManager and Flusher approaches.
Additionally if the ShotoverManager approach doesnt pan out we can always leave it as the Flusher approach.

Comment (conorbros) [reply shotover-proxy/tests/redis_int_tests/basic_driver_tests.rs:1576]: I agree with that. Let's not overengineer it, if we end up needing the abstraction we'll do it.

**Comment by rukai** [comment]
I was reading this article on the weekend, maybe we should use oncecell to store the direct_connection https://matklad.github.io/2022/06/11/caches-in-rust.html

File: shotover-proxy/src/transforms/redis/sink_cluster.rs
Comment (rukai) [inline shotover-proxy/src/transforms/redis/sink_cluster.rs:118]: Im a bit conflicted on if caching the direct connection in self.direct_connection is the right call here.

reasons to cache:
Its more correct as we can maintain connection state for the direct connection for the lifetime of the chain.
Bonus: faster lookup.

reasons not to cache:
maybe we shouldnt be relying on connection state in the first place as that state wont exist when querying a SLOT off of this connection.
bonus: the connection pool will auto-detect and reopen closed connections for us.

At the moment I'm kind of thinking it doesnt matter a whole lot either way right now and we should properly address this when we look into connection state issues like RESP3 support.

File: docs/src/transforms.md
Comment (conorbros) [reply docs/src/transforms.md:283]: I agree, I think these examples should be the most sensible options.

Comment (rukai) [reply docs/src/transforms.md]: alright, done

File: shotover-proxy/example-configs/redis-cluster-hiding/docker-compose.yml
Comment (conorbros) [inline shotover-proxy/example-configs/redis-cluster-hiding/docker-compose.yml:1]: I know you didn't write this whole file, but any idea why this is '2' while the others are '3.3'

Comment (rukai) [reply shotover-proxy/example-configs/redis-cluster-hiding/docker-compose.yml:1]: yeah its a bit all over the place.
I think its out of scope this PR but you could raise an issue to unify onto a single version if you wanted.

File: shotover-proxy/src/transforms/redis/sink_cluster.rs
Comment (benbromhead) [reply shotover-proxy/src/transforms/redis/sink_cluster.rs:118]: I'm ok to revisit this properly when we look into resp3

**Review by conorbros — APPROVED** [review]

**Pull request merged**
